nsis/src/nsis/NsisLoader.java

```diff
@@ -81,56 +91,82 @@ public String getName() {
 	}
 
 	@Override
-	protected void load(ByteProvider provider, LoadSpec loadSpec,
-			List<Option> options, Program program, TaskMonitor monitor,
-			MessageLog log) throws CancelledException, IOException {
-
+	protected void load(ByteProvider provider, LoadSpec loadSpec, List<Option> options,
+			Program program, TaskMonitor monitor, MessageLog log)
+			throws CancelledException, IOException {
 		try {
 			GenericFactory factory = MessageLogContinuesFactory.create(log);
-			NsisExecutable ne = NsisExecutable.createNsisExecutable(factory,
-					provider, SectionLayout.FILE);
-
-			long nsis_header_offset = ne.getHeaderOffset();
-			if (nsis_header_offset == -1) {
-				System.out.print("Could not find nsis_header_offset.\n");
-				return;
+			NsisExecutable ne = NsisExecutable.createNsisExecutable(factory, provider,
+					SectionLayout.FILE);
+			long scriptHeaderOffset = ne.getHeaderOffset();
+			BinaryReader binary_reader = new BinaryReader(provider, /* isLittleEndian= */ true);
+			binary_reader.setPointerIndex(scriptHeaderOffset);
+			Address scriptHeaderAddress = program.getAddressFactory().getDefaultAddressSpace()
+					.getAddress(scriptHeaderOffset);
+
+			FileBytes fileBytesHeader = MemoryBlockUtils.createFileBytes(program, provider,
+					scriptHeaderOffset, NsisScriptHeader.getHeaderSize(), monitor);
+			initScriptHeader(fileBytesHeader, scriptHeaderAddress, fileBytesHeader.getSize(),
+					program, ne.getHeaderDataType());
+
+			FileBytes fileBytesBody;
+			byte[] decompressedBytes = ne.getDecompressedBytes();
+			if (decompressedBytes != null) { // if binary was compressed
+				ByteArrayProvider uncompressedBytes = new ByteArrayProvider(decompressedBytes);
+				fileBytesBody = MemoryBlockUtils.createFileBytes(program, uncompressedBytes, 0,
+						uncompressedBytes.length(), monitor);
+			} else {
+				fileBytesBody = MemoryBlockUtils.createFileBytes(program, provider,
+						scriptHeaderOffset + NsisScriptHeader.getHeaderSize(), ne.getArchiveSize(),
+						monitor);
```

Comment:
Would it be possible to make NsisLoader unaware of compression details of NsisExecutable? To make NsisExecutable expose just a single "getData", which would spit out decompressed bytes if they were compressed or the content at the needed offset.

Reply:
I like the idea! The only downside I see is that for a uncompressed binary, there is extra steps added. If the binary is not compressed, the NsisLoader already has the ByteProvider object to use to import the uncompressed data and does not have to fetch the data. 

Would it be better to change it?

Reply:
I agree with Veranika's suggestion.

I think the handling of the NSIS processing should be put in the `NSISExecutable` class. `NSISExecutable` already creates a `NsisScriptHeader` class. It could also do all the parsing and decompression of the block headers and creating `NSISBlockHeader` objects that already exists. The `NsisLoader` class then is just responsible for taking the data from the `NSISExecutable` and creating the Ghidra objects.

I sort of think of `NsisLoader` as the middleman between `NSISExectuable` and Ghidra. So if Ghidra changes their interface only `NsisLoader` changes.

Reply:
Sounds good! I'll do some modifications with that in mind

Reply:
I would like to address this in a separate pull request as it will require some new code and quite a lot of changes. Issue opened for more details: #19